pd-client/conn.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pd
+
+import (
+	"bufio"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/deadline"
+	"github.com/ngaut/log"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/pd/pkg/apiutil"
+	"github.com/pingcap/pd/pkg/rpcutil"
+)
+
+const (
+	requestPDTimeout   = time.Second
+	connectPDTimeout   = time.Second * 3
+	reconnectPDTimeout = time.Second * 60
+)
+
+const (
+	readBufferSize  = 8 * 1024
+	writeBufferSize = 8 * 1024
+)
+
+type conn struct {
+	net.Conn
+	wg         sync.WaitGroup
+	quit       chan struct{}
+	ConnChan   chan *conn
+	ReadWriter *bufio.ReadWriter
+}
+
+func newConn(c net.Conn) *conn {
+	reader := bufio.NewReaderSize(deadline.NewDeadlineReader(c, requestPDTimeout), readBufferSize)
+	writer := bufio.NewWriterSize(deadline.NewDeadlineWriter(c, requestPDTimeout), writeBufferSize)
+	return &conn{
+		Conn:       c,
+		quit:       make(chan struct{}),
+		ConnChan:   make(chan *conn),
+		ReadWriter: bufio.NewReadWriter(reader, writer),
+	}
+}
+
+func mustNewConn(urls []string, quit chan struct{}) *conn {
+	for {
+		conn, err := rpcConnectLeader(urls)
+		if err == nil {
+			return newConn(conn)
+		}
+		log.Warn(err)
+
+		conn, err = rpcConnect(urls)
+		if err == nil {
+			c := newConn(conn)
+			c.wg.Add(1)
+			go c.connectLeader(urls, reconnectPDTimeout)
```

Comment:
For https://github.com/pingcap/pd/pull/327/files#diff-16d4c46a065cb17906c52afc0c015c9aR127